torch/_inductor/comm_lowering.py

```diff
@@ -0,0 +1,222 @@
+#
+# In collective communication libraries such as NCCL, every rank maintains
+# communication buffers that are remotely accessible by some peers. Depending
+# on the underlying transport, remote accessibility may be established via
+# mechanisms such as ib_reg_mr, CUDA P2P, or CUDA multicast. Typically, these
+# buffers are private to the communication library by default, and
+# communication ops copy user data in and out of these buffers.
+#
+# To prevent these copies, an optimization commonly known as "user buffer
+# registration" can be employed. This allows direct establishment of remote
+# accessibility on user buffers, eliminating the need for copying. However,
+# this optimization introduces stringent usage requirements, which are
+# typically hard to satisfy without being intrusive to the user code:
+#
+# - Establishing remote accessibility is expensive and often done ahead of
+# time. In such implementations, all ranks must agree on the set of allocations
+# used for every collective op. Failing to meet this requirement can
+# lead to runtime errors or even silent correctness issues.
+# - Even if the collective communication library supports gracefully falling
+# back to "unregistered" implementations, the fallback mechanism would nullify
+# the optimization.
+# - Some communication mechanisms impose stricter requirements than others. For
+# example, CUDA's multicast + multi-mem instructions require all ranks to agree
+# not only on the allocations used for every collective but also on the offsets
+# within these allocations.
+#
+# To support all different mechanisms with optimal results, we aim to satisfy
+# the strictest requirement for this family of optimizations - we ensures that
+# every collective op invocation is guaranteed to operate on the same
+# allocation, at the same offset, in every iteration.
+#
+# For eligible collective ops, we identify communication buffers at lowering
+# time and optionally choose to lower the op to a different kernel
+# (ommunication libraries like NCCL handle both registered and non-registered
+# buffers transparently within the same op, though some may require different
+# ops for different cases). Later, the codegen will perform "persistent
+# allocation" to satisfy the aforementioned constraints, and optionally,
+# perform buffer planning to optimize overall memory usage.
+def can_realize_as_comm_buffer(buffer: ir.Buffer, comm_buffer_type: str) -> bool:
+    """
+    Check if a buffer can be realized as the specified `comm_buffer_type`.
+    """
+    assert isinstance(buffer, ir.Buffer), type(buffer)
+
+    # The buffer is already realized as a comm buffer
+    if buffer.get_name() in V.graph.comm_buffer_type:
+        return comm_buffer_type == V.graph.comm_buffer_type[buffer.get_name()]
+
+    # We can realized a buffer as comm buffer only if we know its size at
+    # codegen time.
+    try:
+        V.graph.sizevars.size_hint(buffer.get_numel())
```

Comment:
This seems weird(?). Do you want to check whether it's all static shapes?

Reply:
Yup exactly. Please suggest a better way for doing it

Reply:
We have a `is_symbolic` util: https://github.com/pytorch/pytorch/blob/e18439113e3e9260977e2bc52b83fa83a6fb30b0/torch/_inductor/fx_passes/pad_mm.py#L97

I'm actually not sure how your code even works. Shouldn't we always be able to get a size hint unless it's an unbacked symint?